Support Determining Max Sessions by Authentication

ClaudenirFreitas opened the pull request:
Closes gh-16206

<!--
For Security Vulnerabilities, please use https://pivotal.io/security#reporting
-->

<!--
Before creating new features, we recommend creating an issue to discuss the feature. This ensures that everyone is on the same page before extensive work is done.

Thanks for contributing to Spring Security. Please provide a brief description of your pull-request and reference any related issue numbers (prefix references with gh-).
-->

pivotal-cla commented:
@ClaudenirFreitas Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-security&pullRequestId=16218)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-projects/spring-security&pullRequestId=16218) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

pivotal-cla commented:
@ClaudenirFreitas Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-security&pullRequestId=16218)!

ClaudenirFreitas commented on web/src/main/java/org/springframework/security/web/session/SessionLimitStrategy.java:
```diff
@@ -0,0 +1,51 @@
+ * that there is no limit.
+ *
+ * @author Claudenir Freitas
+ * @since 6.5
```
I am unsure about the version. How to proceed here?

jzheaux replied:
`6.5` is correct since that is the next release that will contain features.

jzheaux commented on config/src/main/java/org/springframework/security/config/annotation/web/configurers/SessionManagementConfigurer.java:
```diff
@@ -588,6 +594,17 @@ private SessionAuthenticationStrategy getSessionAuthenticationStrategy(H http) {
 		return this.sessionAuthenticationStrategy;
 	}
 
+	private SessionLimitStrategy getSessionLimitStrategy() {
```
We can simplify some of this conditional logic by treating `maxSessions` as a `sessionLimitStrategy` internally.

For example `maximumSessions(Integer)`'s implementation could become:

```java
this.sessionLimitStrategy = (authentication) -> maximumSessions;
```

In that case, `this.maximumSessions` would go away.

This is also valuable because it aligns with the common pattern in Spring Security to allow applications to easily override default behavior.

ClaudenirFreitas replied:
Yes, makes total sense. Thanks!

jzheaux commented on web/src/main/java/org/springframework/security/web/authentication/session/ConcurrentSessionControlAuthenticationStrategy.java:
```diff
@@ -172,15 +173,23 @@ public void setExceptionIfMaximumExceeded(boolean exceptionIfMaximumExceeded) {
+	 * unlimited sessions.
+	 * @param sessionLimitStrategy the session limit strategy
+	 */
+	public void setSessionLimitStrategy(SessionLimitStrategy sessionLimitStrategy) {
```
Please also call this `setMaximumSessions` since that is what it's configuring.

jzheaux commented on docs/modules/ROOT/pages/servlet/appendix/namespace/http.adoc:
```diff
@@ -2168,6 +2168,9 @@ Allows injection of the ExpiredSessionStrategy instance used by the ConcurrentSe
 Specify `-1` as the value to support unlimited sessions.
 
+[[nsa-concurrency-control-session-limit-strategy-ref]]
+* **session-limit**
```
Please call this `max-sessions-ref` so as to clarify that it is a strategy bean for determining the session limit.

jzheaux commented on config/src/test/java/org/springframework/security/config/doc/XsdDocumentedTests.java:
```diff
@@ -65,7 +66,7 @@ public class XsdDocumentedTests {
 	String schema31xDocumentLocation = "org/springframework/security/config/spring-security-3.1.xsd";
 
-	String schemaDocumentLocation = "org/springframework/security/config/spring-security-6.4.xsd";
+	String schemaDocumentLocation = "org/springframework/security/config/spring-security-6.5.xsd";
```
There are a few more things that need to be done to update the minor version across the project. Will you please leave these changes out and they will be taken care of by a commit that precedes merging this PR.

ClaudenirFreitas replied:
I updated this one because the [`countReferencesWhenReviewingDocumentationThenEntireSchemaIsIncluded` test](https://github.com/spring-projects/spring-security/blob/15b9a2a4f96bdacbb4ee6050bdbcef5e603bb391/config/src/test/java/org/springframework/security/config/doc/XsdDocumentedTests.java#L181) was failing using the `spring-security-6.4.xsd` file.
Output:
```
...
but could not find the following element(s):
  ["nsa-concurrency-control-max-sessions-ref"]
...
```
The test uses [the `referenceLocation` variable](https://github.com/spring-projects/spring-security/blob/15b9a2a4f96bdacbb4ee6050bdbcef5e603bb391/config/src/test/java/org/springframework/security/config/doc/XsdDocumentedTests.java#L65) which gets [the actual `http.adoc` file](https://github.com/spring-projects/spring-security/blob/15b9a2a4f96bdacbb4ee6050bdbcef5e603bb391/docs/modules/ROOT/pages/servlet/appendix/namespace/http.adoc?plain=1#L2173). 
I suppose we need to keep it, what do you think?

jzheaux replied:
I see. Let's keep it for now so that the build passes. I'll add a note to this ticket to rebase this PR after #16221 is complete.

jzheaux commented on config/src/test/java/org/springframework/security/config/annotation/web/configurers/SessionManagementConfigurerTests.java:
```diff
@@ -625,6 +702,44 @@ UserDetailsService userDetailsService() {
+			// @formatter:off
+			http
+					.formLogin(withDefaults())
+					.sessionManagement((sessionManagement) ->
```
We are moving towards the following syntax:

```java
http
    // ...
    .method((elements) -> elements
        .submethod((sub) -> sub
           // ...
        )
    )
```

That is, the lambda variable name and its usage go on the same line and the end parenthesis goes on the next line, like an end curly brace. Can you update this to use that format please?

jzheaux commented on web/src/main/java/org/springframework/security/web/session/SessionLimitStrategy.java:
```diff
@@ -0,0 +1,51 @@
+ * @since 6.5
+ */
+@FunctionalInterface
+public interface SessionLimitStrategy {
```
Aligning with `SessionLimit` in reactive, will you please have this implement `Function<Authentication, Integer>`?

jzheaux replied:
Then the `resolve` method can be resolved as well.

jzheaux commented on web/src/main/java/org/springframework/security/web/session/SessionLimitStrategy.java:
```diff
@@ -0,0 +1,51 @@
+	 * @return a {@link SessionLimitStrategy} instance that returns the given value.
+	 */
+	static SessionLimitStrategy of(int maxSessions) {
+		Assert.isTrue(maxSessions != 0 && maxSessions >= -1,
```
I think `maxSessions >  0` will simplify this logic, or am I missing something?

ClaudenirFreitas replied:
[`-1` is a valid argument](https://github.com/spring-projects/spring-security/blob/15b9a2a4f96bdacbb4ee6050bdbcef5e603bb391/web/src/main/java/org/springframework/security/web/authentication/session/ConcurrentSessionControlAuthenticationStrategy.java#L131) that defines unlimited sessions.
I also added `maxSessions >= -1` to ensure that we satisfy [this condition](https://github.com/spring-projects/spring-security/blob/15b9a2a4f96bdacbb4ee6050bdbcef5e603bb391/web/src/main/java/org/springframework/security/web/authentication/session/ConcurrentSessionControlAuthenticationStrategy.java#L99). 
One question: the current version allows other negative values and expires all existing sessions. What is the expected behaviour? Should it be kept as is, or should an error be thrown saying the value is invalid?

jzheaux replied:
I see. I would prefer that we abstract away sentinel values. Instead, they should use something like `SessionLimit.UNLIMITED` to indicate that they want unlimited sessions. `of` should only take values greater than 0.

jzheaux commented on config/src/main/java/org/springframework/security/config/http/HttpConfigurationBuilder.java:
```diff
@@ -485,10 +489,16 @@ else if (StringUtils.hasText(sessionAuthStratRef)) {
 			if (StringUtils.hasText(maxSessions)) {
 				concurrentSessionStrategy.addPropertyValue("maximumSessions", maxSessions);
 			}
+			String sessionLimitStrategyRef = this.pc.getReaderContext()
```
We need to error if these two attributes are used in the same declaration as we don't know which one to pick.

ClaudenirFreitas replied:
We already have it. The line `606` is excecuted before this.

jzheaux reviewed (changes requested):
Thanks for such a quick turnaround, @ClaudenirFreitas! I've left my feedback inline.

jzheaux commented on config/src/main/java/org/springframework/security/config/annotation/web/configurers/SessionManagementConfigurer.java:
```diff
@@ -706,7 +709,18 @@ private ConcurrencyControlConfigurer() {
+		 * maximum number of sessions for a user
+		 * @return the {@link ConcurrencyControlConfigurer} for further customizations
+		 */
+		public ConcurrencyControlConfigurer sessionLimitStrategy(SessionLimitStrategy sessionLimitStrategy) {
```
Let's please try and align this with the reactive support and call this method `maximumSessions`.

jzheaux commented on config/src/main/java/org/springframework/security/config/annotation/web/configurers/SessionManagementConfigurer.java:
```diff
@@ -119,12 +120,12 @@ public final class SessionManagementConfigurer<H extends HttpSecurityBuilder<H>>
 
 	private SessionInformationExpiredStrategy expiredSessionStrategy;
 
+	private SessionLimitStrategy sessionLimitStrategy;
```
Let's please try and align this with the reactive support and call it `SessionLimit`.

jzheaux commented on web/src/main/java/org/springframework/security/web/authentication/session/ConcurrentSessionControlAuthenticationStrategy.java:
```diff
@@ -172,15 +173,23 @@ public void setExceptionIfMaximumExceeded(boolean exceptionIfMaximumExceeded) {
+	/**
+	 * Sets the <tt>sessionLimitStrategy</tt> property. The default value is 1. Use -1 for
+	 * unlimited sessions.
+	 * @param sessionLimitStrategy the session limit strategy
```
Please add `@since 6.5`

jzheaux commented on config/src/main/java/org/springframework/security/config/http/HttpConfigurationBuilder.java:
```diff
@@ -485,10 +489,16 @@ else if (StringUtils.hasText(sessionAuthStratRef)) {
+				.getEnvironment()
+				.resolvePlaceholders(sessionCtrlElt.getAttribute(ATT_MAX_SESSIONS_REF));
+			if (StringUtils.hasText(sessionLimitStrategyRef)) {
+				concurrentSessionStrategy.addPropertyReference("sessionLimitStrategy", sessionLimitStrategyRef);
```
I believe this needs to be `maximumSessions` now for the property reference.

I also think it would be clearer if the variable were called `maxSessionsRef` to match the attribute name.

jzheaux commented on config/src/test/resources/org/springframework/security/config/http/HttpHeadersConfigTests-DefaultsSessionManagementConcurrencyControlMaxSessions.xml:
```diff
@@ -0,0 +1,38 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
I imagine you'll add this; however just a reminder to have an XML test that uses `max-sessions-ref`.

ClaudenirFreitas replied:
Tests have been added [here](https://github.com/spring-projects/spring-security/pull/16218/commits/67d40a1c3ac0d84be07d083a81c872133ddffacf).

jzheaux reviewed (changes requested):
Thanks for the updates, @ClaudenirFreitas! I've left some additional feedback inline.

Let me know when you have integration tests ready, and I'll take another look.

ClaudenirFreitas commented on config/src/main/java/org/springframework/security/config/annotation/web/configurers/SessionManagementConfigurer.java:
```diff
@@ -570,7 +571,9 @@ private SessionAuthenticationStrategy getSessionAuthenticationStrategy(H http) {
 			ConcurrentSessionControlAuthenticationStrategy concurrentSessionControlStrategy = new ConcurrentSessionControlAuthenticationStrategy(
 					sessionRegistry);
-			concurrentSessionControlStrategy.setMaximumSessions(this.maximumSessions);
+			if (this.sessionLimitStrategy != null) {
```
The default value is `1` (refer to [this](https://github.com/spring-projects/spring-security/blob/main/web/src/main/java/org/springframework/security/web/authentication/session/ConcurrentSessionControlAuthenticationStrategy.java#L79)). 
Should we add another default value in the `SessionManagementConfigurer` class? (I think we could have only one default value, what do you think?)

jzheaux replied:
Good point. I think the issue here is that the code is trying to use the field for two different things; determine if concurrency control is enabled and determine the maximum number of sessions.

It may work better to separate the two like so:

```java
private SessionLimitStrategy maximumSessions = SessionLimitStrategy.of(1)

private boolean concurrentControlEnabled;
```

ClaudenirFreitas replied:
Checking this, we do not need to the `if (this.sessionLimitStrategy != null) {` because the `isConcurrentSessionControlEnabled` method already checks it three line above, so the `this.sessionLimitStrategy` value will never be null.
With this, we are keeping the same behaviour, is it ok?

ClaudenirFreitas commented:
> Thanks for the updates, @ClaudenirFreitas! I've left some additional feedback inline.
> 
> Let me know when you have integration tests ready, and I'll take another look.

Done.

Many tests failed due to the new release (6.5). I've updated many files on [this commit](https://github.com/spring-projects/spring-security/pull/16218/commits/67d40a1c3ac0d84be07d083a81c872133ddffacf), and the tests are passing now.

Silly question, is this flow correct for the new release?

ClaudenirFreitas commented:
> > Thanks for the updates, @ClaudenirFreitas! I've left some additional feedback inline.
> > Let me know when you have integration tests ready, and I'll take another look.
> 
> @jzheaux integration tests have been added.
> 
> Many tests failed due to the new release (6.5). I've updated many files on [this commit](https://github.com/spring-projects/spring-security/pull/16218/commits/67d40a1c3ac0d84be07d083a81c872133ddffacf), and the tests are passing now.
> 
> Silly question, is this flow correct for the new release?



> > Thanks for the updates, @ClaudenirFreitas! I've left some additional feedback inline.
> > Let me know when you have integration tests ready, and I'll take another look.
> 
> @jzheaux integration tests have been added.
> 
> Many tests failed due to the new release (6.5). I've updated many files on [this commit](https://github.com/spring-projects/spring-security/pull/16218/commits/67d40a1c3ac0d84be07d083a81c872133ddffacf), and the tests are passing now.
> 
> Silly question, is this flow correct for the new release?

@jzheaux 
I just rebased this PR due to the current main changes.
I also added the unit and integration tests. 
Could you review them again?

jzheaux commented on config/src/main/java/org/springframework/security/config/annotation/web/configurers/SessionManagementConfigurer.java:
```diff
@@ -123,7 +124,7 @@ public final class SessionManagementConfigurer<H extends HttpSecurityBuilder<H>>
 	private SessionRegistry sessionRegistry;
 
-	private Integer maximumSessions;
+	private SessionLimitStrategy sessionLimit;
```
Please call the class `SessionLimit` to align with the Reactive name

ClaudenirFreitas replied:
done.

jzheaux commented on config/src/main/java/org/springframework/security/config/annotation/web/configurers/SessionManagementConfigurer.java:
```diff
@@ -706,7 +707,18 @@ private ConcurrencyControlConfigurer() {
+		 * maximum number of sessions for a user
+		 * @return the {@link ConcurrencyControlConfigurer} for further customizations
+		 */
+		public ConcurrencyControlConfigurer maximumSessions(SessionLimitStrategy sessionLimitStrategy) {
```
Please add `@since 6.5` to this method

ClaudenirFreitas replied:
done.

jzheaux reviewed (changes requested):
Just the finishing touches now! Please see my inline feedback.

Also, please squash your commits at this point and format it like so:

```
Add Max Sessions by Authentication

Closes gh-16206
```

ClaudenirFreitas commented:
> Just the finishing touches now! Please see my inline feedback.
> 
> Also, please squash your commits at this point and format it like so:
> 
> ```
> Add Max Sessions by Authentication
> 
> Closes gh-16206
> ```

@jzheaux
I had some conflicts when squashing the commits. Could you squash it before the merge? e.g.


![Screenshot 2024-12-14 at 21 47 39](https://github.com/user-attachments/assets/fc39afee-0827-46df-92ad-a9c0b038556c)

jzheaux commented:
Thanks for asking. We don't prefer to add a merge commit to the history in this case. Do you need help squashing?

ClaudenirFreitas commented:
> Thanks for asking. We don't prefer to add a merge commit to the history in this case. Do you need help squashing?

Yes, please.

jzheaux commented:
Thanks for all your work on this PR, @ClaudenirFreitas! This is now merged into `main`. I also added a polish in 1104b45832a16f6a9b5c4ad818e893350aec0633 to add Kotlin support, adjust `SessionLimit`'s package, and add documentation.

ClaudenirFreitas commented:
> Thanks for all your work on this PR, @ClaudenirFreitas! This is now merged into `main`. I also added a polish in [1104b45](https://github.com/spring-projects/spring-security/commit/1104b45832a16f6a9b5c4ad818e893350aec0633) to add Kotlin support, adjust `SessionLimit`'s package, and add documentation.

Thanks a lot for the help, @jzheaux!

We could add a sample using this config in [the `spring-security-samples` repo](https://github.com/spring-projects/spring-security-samples), what do you think?
